File: backend/src/main/java/com/coing/global/security/handler/OAuth2LoginFailureHandler.java
package com.coing.global.security.handler;

import java.io.IOException;

import com.coing.util.MessageUtil;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.authentication.AuthenticationFailureHandler;
import org.springframework.stereotype.Component;
import org.springframework.web.util.UriComponentsBuilder;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;

@Component
@RequiredArgsConstructor
public class OAuth2LoginFailureHandler implements AuthenticationFailureHandler {

	@Value("${next.server.url}")
	private String frontUrl;

	private final MessageUtil messageUtil;

	@Override
	public void onAuthenticationFailure(HttpServletRequest request, HttpServletResponse response,
		AuthenticationException exception) throws IOException {

		String redirectUrl = UriComponentsBuilder.fromUriString(frontUrl + "/user/social-login")
			.queryParam("error", messageUtil.resolveMessage("login.failure"))
			.toUriString();

		response.sendRedirect(redirectUrl);
	}
}

File: backend/src/main/java/com/coing/global/security/handler/OAuth2LoginFailureHandler.kt
package com.coing.global.security.handler

import com.coing.util.MessageUtil
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.authentication.AuthenticationFailureHandler
import org.springframework.stereotype.Component
import org.springframework.web.util.UriComponentsBuilder
import java.io.IOException

@Component
class OAuth2LoginFailureHandler(
	private val messageUtil: MessageUtil
) : AuthenticationFailureHandler {

	@Value("\${next.server.url}")
	private lateinit var frontUrl: String

	@Throws(IOException::class)
	override fun onAuthenticationFailure(
		request: HttpServletRequest,
		response: HttpServletResponse,
		exception: AuthenticationException
	) {
		val redirectUrl = UriComponentsBuilder
			.fromUriString("$frontUrl/user/social-login")
			.queryParam("error", messageUtil.resolveMessage("login.failure"))
			.toUriString()

		response.sendRedirect(redirectUrl)
	}
}
